lib/msf/core/rpc/v10/rpc_session.rb

```diff
@@ -477,36 +526,46 @@ def rpc_compatible_modules( sid)
     { "modules" => ret }
   end
 
-private
+  private
 
-  def _find_module(mtype,mname)
-    mod = self.framework.modules.create(mname)
-    if(not mod)
-      error(500, "Invalid Module")
-    end
+  INTERACTIVE_OPERATION_SESSIONS = %w[
+    meterpreter
+    mssql
+    postgresql
+    mysql
+    smb
+  ].freeze
+
+  def _find_module(_mtype, mname)
+    mod = framework.modules.create(mname)
+    error(500, 'Invalid Module') if mod.nil?
 
     mod
   end
 
-  def _valid_session(sid,type)
-
-    s = self.framework.sessions[sid.to_i]
-
-    if(not s)
-      error(500, "Unknown Session ID #{sid}")
-    end
-
-    if type == "ring"
-      if not s.respond_to?(:ring)
-        error(500, "Session #{s.type} does not support ring operations")
+  def _valid_session(sid, operation_type)
+    session = framework.sessions[sid.to_i]
+    error(500, "Unknown Session ID #{sid}") if session.nil?
+
+    case operation_type
+    when 'ring'
+      error(500, "Session of #{session.type} type does not support ring operations") unless session.respond_to?(:ring)
+    when 'meterpreter'
+      error(500, 'Invalid session type') unless session.type == 'meterpreter'
+    when 'interactive'
+      unless INTERACTIVE_OPERATION_SESSIONS.include?(session.type)
```

Comment:
Not an issue with this PR but thinking more broadly should our sessions have a common interface so we don't need to do this check in pro? seems a bit awkward to have to update this whenever we want to add new session support into pro

Reply:
I think it's good improvement for future, new sessions are definitely going to support "interactive" mechanism or whatever we call it, so we would just need to update some of the legacy sessions to eliminate the extra reading method